pilot/pkg/networking/plugin/authn/authentication.go

```diff
@@ -314,3 +328,39 @@ func isDestinationExcludedForMTLS(destService string, mtlsExcludedServices []str
 	}
 	return false
 }
+
+func getJwksURI(policyJwt *authn.Jwt) (string, error) {
+	// Return directly if policyJwt.JwksUri is explicitly set, ex, when policyJwt.Issuer is an email address.
+	if policyJwt.JwksUri != "" {
+		return policyJwt.JwksUri, nil
+	}
+
+	// If policyJwt.Issuer isn't set, try to get it through OpenID Discovery.
+	discoveryURL := policyJwt.Issuer + openIDDiscoveryCfgURLSuffix
+	resp, err := http.Get(discoveryURL)
+	if err != nil {
+		return "", err
+	}
+	defer func() {
+		if errr := resp.Body.Close(); errr != nil {
+			log.Errorf("Failed to close web response: %v", errr)
+		}
+	}()
+
+	body, err := ioutil.ReadAll(resp.Body)
```

Comment:
Is ioutil.ReadAll a blocking call? Can we specify a timeout for it? We do not want to block forever if the server is not responsive. Not all providers support OpenID discovery. For some providers, you may receive 404 Not Found. In this case, will the error appear in "resp" or "err"?

Reply:
There is no timeout setting according to https://golang.org/pkg/io/ioutil/#ReadAll; for 404 case, we still can get response back then hit error later in json.Unmarshal

Reply:
You can follow this example to set timeout  https://github.com/istio/istio/blob/master/mixer/test/client/env/http_client.go#L39

Reply:
Thanks, done.